auth/standard_role_manager.cc

```diff
@@ -250,26 +251,27 @@ future<> standard_role_manager::start() {
             }
             return make_ready_future<>();
         }).then([this] {
-            _stopped = auth::do_after_system_ready(_as, [this] {
-                return seastar::async([this] {
-                    if (legacy_mode(_qp)) {
-                        _migration_manager.wait_for_schema_agreement(_qp.db().real_database(), db::timeout_clock::time_point::max(), &_as).get();
-
-                        if (any_nondefault_role_row_satisfies(_qp, &has_can_login).get()) {
-                            if (legacy_metadata_exists()) {
-                                log.warn("Ignoring legacy user metadata since nondefault roles already exist.");
-                            }
-
-                            return;
-                        }
+            _stopped = auth::do_after_system_ready(_as, [this] () -> future<>  {
+                if (legacy_mode(_qp)) {
+                    co_await _migration_manager.wait_for_schema_agreement(_qp.db().real_database(), db::timeout_clock::time_point::max(), &_as);
 
+                    if (co_await any_nondefault_role_row_satisfies(_qp, &has_can_login)) {
                         if (legacy_metadata_exists()) {
-                            migrate_legacy_metadata().get();
-                            return;
+                            log.warn("Ignoring legacy user metadata since nondefault roles already exist.");
                         }
+
+                        _superuser_created.signal();
+                        co_return;
                     }
-                    create_default_role_if_missing().get();
-                });
+
+                    if (legacy_metadata_exists()) {
+                        co_await migrate_legacy_metadata();
+                        _superuser_created.signal();
+                        co_return;
+                    }
+                }
+                co_await create_default_role_if_missing();
+                _superuser_created.signal();
```

Comment:
maybe single defer() in the right scope would be better?

like defer([](){ _superuser_created.signal()}

Reply:
also btw not sure what happens if exception is thrown during superuser creation, worth to make sure startup doesn't hang, it should terminate.

perhaps some glue with semaphore::broken is needed

Reply:
yes